src/hooks/usePermit2.ts

```diff
@@ -0,0 +1,75 @@
+import { PERMIT2_ADDRESS } from '@uniswap/permit2-sdk'
+import { CurrencyAmount, Token } from '@uniswap/sdk-core'
+import { useWeb3React } from '@web3-react/core'
+import ms from 'ms.macro'
+import { useCallback, useEffect, useMemo, useState } from 'react'
+
+import { ApprovalTransactionInfo } from '..'
+import useInterval from './useInterval'
+import { PermitSignature, usePermitAllowance, usePermitAllowanceCallback } from './usePermitAllowance'
+import { useTokenAllowance, useTokenAllowanceCallback } from './useTokenAllowance'
+
+export enum PermitState {
+  UNKNOWN,
+  APPROVAL_NEEDED,
+  PERMIT_NEEDED,
+  PERMITTED,
+}
+
+export interface Permit {
+  state: PermitState
+  signature?: PermitSignature
+  callback?: () => Promise<ApprovalTransactionInfo | void>
+}
+
+export default function usePermit(amount?: CurrencyAmount<Token>, spender?: string): Permit {
+  const { account } = useWeb3React()
+  const tokenAllowance = useTokenAllowance(amount?.currency, account, PERMIT2_ADDRESS)
+  const approveToken = useTokenAllowanceCallback(amount, PERMIT2_ADDRESS)
+
+  const allowanceData = usePermitAllowance(amount?.currency, spender)
+  const [permitAllowance, setPermitAllowance] = useState(allowanceData?.amount)
+  useEffect(() => setPermitAllowance(allowanceData?.amount), [allowanceData?.amount])
+
+  const [signature, setSignature] = useState<PermitSignature>()
+  const permitToken = usePermitAllowanceCallback(amount?.currency, spender, allowanceData?.nonce, setSignature)
+
+  const approveAndPermitToken = useCallback(async () => {
+    // Queue both transactions. Delay the permit to ensure the approval is prompted for first.
+    const info = approveToken()
+    await new Promise((resolve) => setTimeout(resolve, 500))
+    await permitToken()
```

Comment:
it looks like `approveToken` is a promise. Could you `.then` chain or `await` the call to trigger `permitToken` instead of using the timeout?